src/ott/solvers/quadratic/histogram_transport.py

```diff
@@ -0,0 +1,243 @@
+class HistogramTransport:
+  """Histogram Transport solver.
+
+  Computes the First Lower Bound distance between two distributions.
+  The current implementation requires uniform marginals.
+  If there are an uneven number of points in the two distributions,
+  then we perform a stratified subsample of the distribution of
+  distances to be able to approximate the Wasserstein distance between
+  the local distributions of distnaces.
+
+  Args:
+  softness: regularization for soft sort. Set to `-1` for normal sorting
+  p: exponent for computing the transport distance betweeen histograms
+  epsilon: regularization parameter for the resulting Sinkhorn problem
+  min_iterations: minimum iterations for computing Sinkhorn distance
+  max_iterations: maximum iterations for computing Sinkhorn distance
+  """
+
+  def __init__(
+      self,
+      softness: float = -1.0,
+      p: float = 1.0,
+      epsilon: float = 1.0,
+      min_iterations: int = 10,
+      max_iterations: int = 100,
+  ):
+    self.softness = softness
+    self.p = p
+    self.epsilon = epsilon
+    self.linear_ot_solver = sinkhorn.Sinkhorn(
+        max_iterations=max_iterations, min_iterations=min_iterations
+    )
+
+  def __call__(
+      self,
+      prob: quadratic_problem.QuadraticProblem,
+      rng: Optional[jax.random.PRNGKeyArray] = None,
+  ) -> HTOutput:
+    """Run the Histogram Transport solver.
+
+    Args:
+    prob: Quadratic OT problem.
+    rng: Random number key.
+
+    Returns:
+    The Histogram Transport output.
+    """
+    rng = utils.default_prng_key(rng)
+
+    dists_xx = prob.geom_xx.cost_matrix
+    dists_yy = prob.geom_yy.cost_matrix
+    m, n = dists_xx.shape[0], dists_yy.shape[0]
+    min_num_pts = min([m, n])
+    small_indices = jnp.arange(min_num_pts)
+    x_indices = jnp.round(small_indices * (m / min_num_pts)).astype(int)
+    y_indices = jnp.round(small_indices * (n / min_num_pts)).astype(int)
+
+    if self.softness < 0:
+      sorted_dists_xx = jax.lax.sort(dists_xx, dimension=-1)
+      sorted_dists_yy = jax.lax.sort(dists_yy, dimension=-1)
+    else:
+      sorted_dists_xx = soft_sort.sort(dists_xx, axis=-1, epsilon=self.softness)
+      sorted_dists_yy = soft_sort.sort(dists_yy, axis=-1, epsilon=self.softness)
+
+    # Uniformly subsample distances
+    sorted_dists_xx = jnp.take_along_axis(
+        sorted_dists_xx, x_indices[None, :], -1
+    )
+    sorted_dists_yy = jnp.take_along_axis(
+        sorted_dists_yy, y_indices[None, :], -1
+    )
+
+    match self.p:
+      case 1.0:
+        cost_xy = jnp.sum(
+            jnp.abs(sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :]),
+            axis=-1,
+        )
+      case 2.0:
+        cost_xy = jax.lax.sqrt(
+            jnp.sum(
+                jnp.square(
+                    sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :]
+                ),
+                axis=-1,
+            )
+        )
+      case _:
+        cost_xy = jnp.power(
```

Comment:
I think 1D transport should be implemented in `solvers.linear`, largely by taking what you have implemented above, and then vmapped here to obtain `cost_xy`. This might then, for instance, be reused for sliced implementation.

Reply:
Good idea! I put it in `wasserstein_1d.py`--might make sense to rename it, though.